lte/cloud/go/services/subscriberdb/servicers/protected/indexer_servicer_test.go

```diff
@@ -96,3 +100,128 @@ func encodeIP(ip string) string {
+
+	// two gatewayIDs
+	serializedGatewayState = state_types.SerializedStatesByID{
+		id0: {SerializedReportedState: serialize(t, &gatewayState1, lte.GatewaySubscriberStateType)},
+		id1: {SerializedReportedState: serialize(t, &gatewayState0, lte.GatewaySubscriberStateType)},
+	}
+	errs, err = idx.Index("nid0", serializedGatewayState)
+	assert.NoError(t, err)
+	assert.Empty(t, errs)
+	foundStates, err = subscriberStorage.GetSubscribersForGateway("nid0", gwid1)
+	assert.NoError(t, err)
+	assert.True(t, cmp.Equal(*foundStates, gatewayState1))
+	foundStates, err = subscriberStorage.GetSubscribersForGateway("nid0", gwid2)
+	assert.NoError(t, err)
+	assert.True(t, cmp.Equal(*foundStates, gatewayState0))
+
+	// mixed state (two gateways, two mobilityd states)
+	// define mobilityd state ids
+	idm0 := state_types.ID{Type: lte.MobilitydStateType, DeviceID: "IMSI0.apn0"}
+	idm1 := state_types.ID{Type: lte.MobilitydStateType, DeviceID: "IMSI0.apn1"}
+
+	states := state_types.SerializedStatesByID{
+		id0: {SerializedReportedState: serialize(t, &gatewayState0, lte.GatewaySubscriberStateType)},
+		idm0: {SerializedReportedState: serialize(t, &state.ArbitraryJSON{"ip": state.ArbitraryJSON{"address": encodeIP("127.0.0.1")}}, lte.MobilitydStateType)},
+		idm1: {SerializedReportedState: serialize(t, &state.ArbitraryJSON{"ip": state.ArbitraryJSON{"address": encodeIP("127.0.0.1")}}, lte.MobilitydStateType)},
+		id1: {SerializedReportedState: serialize(t, &gatewayState0, lte.GatewaySubscriberStateType)},
+	}
+	errs, err = idx.Index("nid0", states)
+	assert.NoError(t, err)
+	assert.Empty(t, errs)
+	foundStates, err = subscriberStorage.GetSubscribersForGateway("nid0", gwid1)
+	assert.NoError(t, err)
+	assert.True(t, cmp.Equal(*foundStates, gatewayState0))
+	foundStates, err = subscriberStorage.GetSubscribersForGateway("nid0", gwid2)
+	assert.NoError(t, err)
+	assert.True(t, cmp.Equal(*foundStates, gatewayState0))
+	foundIPs, err := subscriberdb.GetIMSIsForIP(context.Background(), "nid0", "127.0.0.1")
+	assert.NoError(t, err)
+	assert.Equal(t, []string{"IMSI0"}, foundIPs)
+
+	// empty gateways
+	emptySerializedState = state_types.SerializedStatesByID{
+		id0: {SerializedReportedState: serialize(t, &emptyState, lte.GatewaySubscriberStateType)},
+		id1: {SerializedReportedState: serialize(t, &emptyState, lte.GatewaySubscriberStateType)},
+	}
+	errs, err = idx.Index("nid0", emptySerializedState)
+	assert.NoError(t, err)
+	assert.Empty(t, errs)
+	foundStates, err = subscriberStorage.GetSubscribersForGateway("nid0", gwid1)
+	assert.NoError(t, err)
+	assert.True(t, cmp.Equal(*foundStates, emptyState))
+	foundStates, err = subscriberStorage.GetSubscribersForGateway("nid0", gwid2)
+	assert.NoError(t, err)
+	assert.True(t, cmp.Equal(*foundStates, emptyState))
+
+	// two identical gateway IDs: Shouldn't happen. but are we protected against that?
+
+}
+
+
+func CreateTestGatewaySubscriberState(imsis ...string) storage.GatewaySubscriberState {
```

Comment:
This is a copy and still needs a better place. One can't import functions from _test.go files.
(Tried: test_utils.go. Problem: When putting this function into a test_utils.go one has to import storage.GatewaySubscriberState and in subscriber_storage_test.go one has to import test_utils.go, which results in a circle that is not allowed.)

Reply:
I don't see how that leads to a circle. Both files import GatewaySubscriberState but that shouldn't be a problem?

Reply:
I think the problem arises since the subscriber_storage_test.go file also belongs to the storage package. Then it goes like that:

package test_utils (test_utils.go) --> package storage (subscriber_state.go/GatewaySubscriberState)
package storage (subscriber_state_test.go) --> package test_utils (test_utils.go/Create...)

Reply:
Can the file test_utils also go into the package storage?

Reply:
Ah, sure. That should work.